libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c

```diff
@@ -355,17 +360,111 @@ AwsIotProvisioningError_t _AwsIotProvisioning_SerializeCreateKeysAndCertificateR
 
 /*------------------------------------------------------------------*/
 
-bool _AwsIotProvisioning_SerializeCreateCertificateFromCsrRequestPayload( const char * pCertificateSigningRequest,
-                                                                          size_t csrLength,
-                                                                          uint8_t * pSerializationBuffer,
-                                                                          size_t * pBufferSize )
+bool _AwsIotProvisioning_SerializeCreateCertFromCsrRequestPayload( const char * pCertificateSigningRequest,
```

Comment:
This function is over-complicated because it has two different jobs.  It has the job described by the name of the function, which is to serialize a create certificate request from the CSR payload, but it has a second job which is to determine how much buffer space is required to do the first.  Because we have these two jobs, we end up with 3 different places where we're checking to see which job we're doing: line 374, 389, and line 454.

To fix this, create a `static` helper function for job 2 (calculating the size required).  That way, you'll only need to test `pSerializationBuffer == NULL` once at line 374 and this function will be a lot simpler and easier to read.

Reply:
Yes, good suggestion. I will break this function into 2 for better readability!